ddtrace/llmobs/_integrations/langgraph.py

```diff
@@ -0,0 +1,185 @@
+
+    def _llmobs_set_tags(
+        self,
+        span: Span,
+        args: List[Any],
+        kwargs: Dict[str, Any],
+        response: Optional[Any] = None,
+        operation: str = "",  # oneof graph, node
+    ):
+        if not self.llmobs_enabled:
+            return
+
+        inputs = get_argument_value(args, kwargs, 0, "input")
+        span_name = kwargs.get("name", span.name)
+
+        span._set_ctx_items(
+            {
+                SPAN_KIND: "agent",  # should nodes be workflows? should it be dynamic to if a subgraph is included?
+                INPUT_VALUE: format_langchain_io(inputs),
+                OUTPUT_VALUE: format_langchain_io(response),
+                NAME: span_name,
+            }
+        )
+
+        if operation != "node":
+            return  # we set the graph span links in handle_pregel_loop_tick
+
+        config = get_argument_value(args, kwargs, 1, "config")
+
+        metadata = config.get("metadata", {}) if isinstance(config, dict) else {}
+        node_instance_id = metadata["langgraph_checkpoint_ns"].split(":")[-1]
+
+        node_invoke = node_invokes[node_instance_id] = node_invokes.get(node_instance_id, {})
+        node_invoke["span"] = {
+            "trace_id": "{:x}".format(span.trace_id),
+            "span_id": str(span.span_id),
+        }
+
+        node_invoke_span_links = node_invoke.get("from")
+
+        span_links = (
+            [
+                {
+                    "span_id": str(_get_llmobs_parent_id(span)) or "undefined",
+                    "trace_id": "{:x}".format(span.trace_id),
+                    # we assume no span link means it is the first node of a graph
+                    "attributes": {
+                        "from": "input",
+                        "to": "input",
+                    },
+                }
+            ]
+            if node_invoke_span_links is None
+            else node_invoke_span_links
+        )
+
+        current_span_links = span._get_ctx_item(SPAN_LINKS) or []
+        span._set_ctx_item(SPAN_LINKS, current_span_links + span_links)
+
+    def handle_pregel_loop_tick(self, finished_tasks: dict, next_tasks: dict, more_tasks: bool):
+        """
+        Handle a specific tick of the pregel loop.
+        Specifically, this function computes incoming and outgoing span links between finished tasks
+        and queued tasks in the graph.
+
+        Additionally, it sets the span links at the outer ends of the graph, between the span that invokes
+        the graph and the last set of nodes before the graph ends.
+        """
+        if not self.llmobs_enabled:
+            return
+
+        graph_span = (
+            tracer.current_span()
+        )  # since we're running the the pregel loop, and not in a node, the graph span should be the current span
+        graph_caller = _get_nearest_llmobs_ancestor(graph_span) if graph_span else None
+
+        if not more_tasks and graph_span is not None:
```

Comment:
The graph is finished, we are now setting span links from the overarching graph span to the outputs of this graph.